sphinx/pycode/ast.py

```diff
@@ -64,113 +64,137 @@ def unparse(node: ast.AST) -> str:
+
+    def visit_Bytes(self, node):
         return repr(node.s)
-    elif isinstance(node, ast.Call):
-        args = ([unparse(e) for e in node.args] +
-                ["%s=%s" % (k.arg, unparse(k.value)) for k in node.keywords])
-        return "%s(%s)" % (unparse(node.func), ", ".join(args))
-    elif isinstance(node, ast.Dict):
-        keys = (unparse(k) for k in node.keys)
-        values = (unparse(v) for v in node.values)
+
+    def visit_Call(self, node):
+        args = ([self.visit(e) for e in node.args] +
+                ["%s=%s" % (k.arg, self.visit(k.value)) for k in node.keywords])
+        return "%s(%s)" % (self.visit(node.func), ", ".join(args))
+
+    def visit_Dict(self, node):
+        keys = (self.visit(k) for k in node.keys)
+        values = (self.visit(v) for v in node.values)
         items = (k + ": " + v for k, v in zip(keys, values))
         return "{" + ", ".join(items) + "}"
-    elif isinstance(node, ast.Ellipsis):
+
+    def visit_Ellipsis(self, node):
         return "..."
-    elif isinstance(node, ast.Index):
-        return unparse(node.value)
-    elif isinstance(node, ast.Lambda):
-        return "lambda %s: ..." % unparse(node.args)
-    elif isinstance(node, ast.List):
-        return "[" + ", ".join(unparse(e) for e in node.elts) + "]"
-    elif isinstance(node, ast.Name):
-        return node.id
-    elif isinstance(node, ast.NameConstant):
-        return repr(node.value)
-    elif isinstance(node, ast.Num):
-        return repr(node.n)
-    elif isinstance(node, ast.Set):
-        return "{" + ", ".join(unparse(e) for e in node.elts) + "}"
-    elif isinstance(node, ast.Str):
-        return repr(node.s)
-    elif isinstance(node, ast.Subscript):
-        return "%s[%s]" % (unparse(node.value), unparse(node.slice))
-    elif isinstance(node, ast.UnaryOp):
-        return "%s %s" % (unparse(node.op), unparse(node.operand))
-    elif isinstance(node, ast.Tuple):
-        if node.elts:
-            return ", ".join(unparse(e) for e in node.elts)
-        else:
-            return "()"
-    elif sys.version_info > (3, 6) and isinstance(node, ast.Constant):
-        # this branch should be placed at last
-        return repr(node.value)
-    else:
-        raise NotImplementedError('Unable to parse %s object' % type(node).__name__)
 
+    def visit_Index(self, node):
+        return self.visit(node.value)
 
-def _unparse_arg(arg: ast.arg, default: Optional[ast.AST]) -> str:
-    """Unparse a single argument to a string."""
-    name = unparse(arg)
-    if default:
-        if arg.annotation:
-            name += " = %s" % unparse(default)
-        else:
-            name += "=%s" % unparse(default)
-    return name
+    def visit_Lambda(self, node):
+        return "lambda %s: ..." % self.visit(node.args)
 
+    def visit_List(self, node):
+        return "[" + ", ".join(self.visit(e) for e in node.elts) + "]"
 
-def unparse_arguments(node: ast.arguments) -> str:
```

Comment:
Diff here is mostly just an extra level of indent, I'd recommend checking the "ignore whitespace" box to review.